### PR TITLE
Collection#set style (Backwards Compatible)

### PR DESCRIPTION
**Pull request by jridgewell**
This is a backwards compatible version of #3648.

- Declares variables where they are needed (and removes duplicate `var i` declarations)
- Exits early when `models` is nullish
- Makes the "merge" and "add" loop easier to follow
- Makes inserting models with `at` an order of magnitude faster
  - [Base case (`#unshift`)](http://jsperf.com/insert-array-into-another-array/14)
  - [Base case (`#push`)](http://jsperf.com/insert-array-into-another-array/15)
  - ["Normal" collection size](http://jsperf.com/insert-array-into-another-array/16)
  - [Large collection](http://jsperf.com/insert-array-into-another-array/17)
  - [Big fetch](http://jsperf.com/insert-array-into-another-array/18)

Depending on how much performance is an issue, we can use an even faster splice algorithm (also included in the jsperfs), it's just a bit complicated.

### CODE REVIEW

File: backbone.js [hunk 1]
```diff
@@ -788,83 +797,88 @@
...
     set: function(models, options) {
+      if (models == null) return;
+
       options = _.defaults({}, options, setOptions);
       if (options.parse && !this._isModel(models)) models = this.parse(models, options);
+
       var singular = !_.isArray(models);
-      models = singular ? (models ? [models] : []) : models.slice();
-      var id, model, attrs, existing, sort;
+      models = singular ? [models] : _.clone(models);
```
Comment (akre54) [inline backbone.js]: is `clone` faster here? Shouldn't it be quicker to avoid the logic branch in clone?

**Comment by akre54** [comment]
:+1:

File: backbone.js (hunk 1 above)
Comment (jridgewell) [inline backbone.js]: Ha, didn't even notice I changed that.

File: test/collection.js [hunk 2]
```diff
@@ -1565,7 +1565,7 @@
     collection.on('sort', function() {
       ok(true);
     });
-    collection.set([{id: 3}, {id: 2}, {id: 1}, {id: 0}]);
+    collection.set([{id: 1}, {id: 2}, {id: 3}, {id: 0}]);
```
Comment (akre54) [inline test/collection.js]: What's the reason for this change?

File: test/collection.js [hunk 3]
```diff
@@ -506,7 +506,7 @@
     };
     collection.url = '/test';
     collection.fetch();
-    this.syncArgs.options.success();
+    this.syncArgs.options.success([]);
```
Comment (akre54) [inline test/collection.js]: What's the reason for this change?

File: test/collection.js (hunk 2 above)
Comment (jridgewell) [inline test/collection.js]: "Adding a model should trigger a sort". Seemed better to _only_ add a model, as the mixed order case is handled in the [test above](https://github.com/jashkenas/backbone/pull/3663/files#diff-e95d8d577a70e10552b2cdfe9a7b1178R1549).

File: test/collection.js (hunk 3 above)
Comment (jridgewell) [inline test/collection.js]: I [preemptively return](https://github.com/jashkenas/backbone/pull/3663/files#diff-0d56d0d310de7ff18b3cef9c2f8f75dcR800) on `null` now, instead of letting it go through parse.

Comment (akre54) [inline test/collection.js]: Ok but that's a breaking change. Any way to still allow `success()` with no args?

Comment (jridgewell) [inline test/collection.js]: We can remove the check, but I don't think `collection.set()` should clear out the collection...

File: test/collection.js (hunk 2 above)
Comment (akre54) [inline test/collection.js]: fair nuf. maybe split this out?

File: test/collection.js (hunk 3 above)
Comment (akre54) [inline test/collection.js]: Agreed. Lets add a test for that.

File: backbone.js [hunk 4]
```diff
@@ -746,6 +746,15 @@
...
   var addOptions = {add: true, remove: false};
 
+  // Splices `insert` into `array` at index `at`.
+  var splice = function(array, insert, at) {
+    var tail = Array(array.length - at);
+    var length = insert.length;
+    for (var i = 0; i < tail.length; i++) tail[i] = array[i + at];
+    for (i = 0; i < length; i++) array[i + at] = insert[i];
+    for (i = 0; i < tail.length; i++) array[i + length + at] = tail[i];
+  };
```
Comment (megawac) [inline backbone.js]: Whats the advantage of this over the Array.prototype.splice

Comment (jridgewell) [inline backbone.js]: Speed. :racehorse: (How is there not a racecar emoji?) See the OP.

Comment (megawac) [inline backbone.js]: Proof of speed?

Comment (jridgewell) [inline backbone.js]: It's in the [OP](https://github.com/jashkenas/backbone/pull/3663#issue-84837714). :stuck_out_tongue:

File: test/collection.js (hunk 3 above)
Comment (jridgewell) [inline test/collection.js]: Added a test.

File: test/collection.js (hunk 2 above)
Comment (jridgewell) [inline test/collection.js]: Split it into what?

**Comment by jashkenas** [comment]
Changing `set` internals is always an adventure — I hope this doesn't break some unexpected things ... but — `splice` function unfortunate black magic aside, this looks pretty nice.
